solutions/gold/izho-17-HardRoute.mdx

```diff
@@ -0,0 +1,159 @@
+---
+id: izho-17-HardRoute
+source: IZhO 2017
+title: Hard Route
+author: Justin Ji
+---
+
+## Explanation
+
+Assume there is some hard route that goes from vertex $u$ to vertex $v$.
+Let the node that the path from $u$ to $v$ and the furthest node from the hard route meet be node $x$.
+Use rerooting DP to calculate the hardest route and the number of such hardest routes for every node $x$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
+## Implementation
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+int main() {
+	int n;
+	cin >> n;
+	vector<vector<int>> adj(n);
+	for (int i = 1; i < n; i++) {
+		int x, y;
+		cin >> x >> y;
+		adj[--x].push_back(--y);
+		adj[y].push_back(x);
+	}
+
+	vector<int> max_length(n);
+	vector<int> path_count(n);
+	/**
+	 * Calculates the longest path from vertex u,
+	 * and the number of such paths.
+	 */
+	function<void(int, int)> dfs = [&](int u, int p) {
+		max_length[u] = 0;
+		path_count[u] = 1;
+		for (int v : adj[u])
+			if (v != p) {
+				dfs(v, u);
+				if (max_length[u] < max_length[v] + 1) {
+					max_length[u] = max_length[v] + 1;
+					path_count[u] = path_count[v];
+				} else if (max_length[v] + 1 == max_length[u]) {
+					path_count[u] += path_count[v];
+				}
+			}
+	};
+
+	dfs(0, -1);
+
+	ll max_hardness = 0;
+	ll hardest_path_count = 1;
+	/**
+	 * Performs the rerooting, to count the hardest
+	 * path and the # of such paths at this vertex.
+	 */
+	function<void(int, int, ll, ll)> dfs2 = [&](int u, int p, ll parDist,
+	                                            ll parCnt) {
```

Comment:
use snake_case for naming? perhaps `parent_dist` and `parent_count` would be great

Reply:
aight bet, give me a sec